File: tests/interop/chrome-redis-safari/create_state.mjs
/**
 * Script to create browser state in Chrome and store in Redis
 * for cross-browser interoperability testing with Safari.
 */

import path from 'path';
import { fileURLToPath } from 'url';
import { BrowserState } from '../../../typescript/dist/index.js';
import { chromium } from 'playwright';

// Get current file directory with ES modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Test constants
const SESSION_ID = 'chrome_to_safari_test';
const USER_ID = 'browser_interop_user';

// Debug mode - set to true to see browser UI
const DEBUG = false;

// Redis configuration 
const REDIS_CONFIG = {
    host: 'localhost',
    port: 6379,
    password: undefined, 
    db: 0,
    keyPrefix: 'browserstate',
    ttl: 604800  // 7 days
};

// Path to the test HTML file
const TEST_HTML_PATH = path.resolve(path.join(__dirname, '../../../typescript/examples/shared/test.html'));
const TEST_URL = `file://${TEST_HTML_PATH}`;

// Function to fail the test with a clear error message
function failTest(message) {
    console.error(`\n❌ TEST FAILED: ${message}`);
    process.exit(1);
}

async function createTestData() {
    console.log('🚀 Starting Chrome State Creation\n');

    try {
        // Initialize BrowserState with Redis storage
        console.log('🔧 Creating BrowserState with Redis storage...');
        const browserState = new BrowserState({
            userId: USER_ID,
            storageType: 'redis',
            redisOptions: REDIS_CONFIG
        });

        // List existing sessions
        console.log('\n📋 Listing existing sessions...');
        const sessions = await browserState.listSessions();
        console.log(`Found ${sessions.length} session(s): ${sessions.join(', ')}`);

        // Mount the session
        console.log(`\n📥 Mounting session: ${SESSION_ID}`);
        const userDataDir = await browserState.mount(SESSION_ID);
        console.log(`📂 Mounted at: ${userDataDir}`);

        // Launch Chrome browser with the mounted state
        console.log('\n🌐 Launching Chrome browser with Playwright...');
        const browser = await chromium.launchPersistentContext(userDataDir, {
            headless: !DEBUG
        });

        try {
            // Create a new page
            const page = await browser.newPage();

            // Navigate to the test HTML page
            console.log(`\n📄 Loading test page: ${TEST_URL}`);
            await page.goto(TEST_URL);
            console.log('✅ Test page loaded');
            
            // Clear existing localStorage to start fresh
            await page.evaluate(() => {
                localStorage.clear();
            });
            
            // Wait for the page to load completely
            await page.waitForTimeout(1000);

            // Add Chrome-specific test data to localStorage
            console.log('\n📝 Adding Chrome-specific test data to localStorage...');
            
            // Add notes to localStorage
            const testNotes = [
                'Chrome created note 1',
                'Chrome created note 2',
                'Chrome created note 3'
            ];

            for (const note of testNotes) {
                await page.fill('#noteInput', note);
                await page.click('button:text("Add Note")');
                await page.waitForTimeout(500); // Wait for animation
            }

            // Add browser metadata to localStorage
            await page.evaluate(() => {
                localStorage.setItem('browserMetadata', JSON.stringify({
                    browser: 'Chrome',
                    timestamp: new Date().toISOString(),
                    userAgent: navigator.userAgent
                }));
            });

            // Verify notes were added
            const notesCount = await page.evaluate(() => {
                return JSON.parse(localStorage.getItem('notes') || '[]').length;
            });
            
            // Verify notes were added successfully
            if (notesCount !== testNotes.length) {
                failTest(`Expected ${testNotes.length} notes, but found ${notesCount}`);
            }
            
            console.log(`✅ Added ${notesCount} notes`);

            // Get the notes data for verification
            const notesData = await page.evaluate(() => {
                return localStorage.getItem('notes');
            });
            
            // Get browser metadata
            const metadataJson = await page.evaluate(() => {
                return localStorage.getItem('browserMetadata');
            });
            
            // Verify data is not empty
            if (!notesData || !metadataJson) {
                failTest('Test data is empty');
            }
            
            console.log(`📝 Notes data: ${notesData}`);
            console.log(`📝 Browser metadata: ${metadataJson}`);
            
            // Wait to ensure data is properly saved
            await page.waitForTimeout(1000);

        } finally {
            await browser.close();
        }

        // Unmount the session
        console.log('\n🔒 Unmounting session...');
        await browserState.unmount();
        console.log('✅ Session unmounted');

        console.log('\n✨ Chrome state creation complete!');
        console.log('Now run verify_state.mjs to verify the state in Safari');

    } catch (error) {
        failTest(`Error during Chrome state creation: ${error instanceof Error ? error.message : String(error)}`);
    }
}

// Run the state creation
createTestData(); 

File: tests/interop/python-chrome-redis-safari/create_state.py
#!/usr/bin/env python3
"""
Python to Safari Browser Interoperability Test

This script creates browser state in Chrome and stores it in Redis,
which can then be loaded by Safari to test cross-browser interoperability.
"""

import os
import json
import sys
import asyncio
from pathlib import Path
from playwright.async_api import async_playwright

# Try different import patterns to handle potential module structure issues
try:
    from browserstate import BrowserState, BrowserStateOptions, RedisStorage
    print("✅ Successfully imported browserstate from top-level package")
except ImportError:
    try:
        # Try importing from submodules
        from browserstate.browser_state import BrowserState, BrowserStateOptions
        from browserstate.storage.redis_storage import RedisStorage
        print("✅ Successfully imported browserstate from submodules")
    except ImportError as e:
        print(f"❌ Failed to import browserstate: {e}")
        print("System path:", sys.path)
        sys.exit(1)

# Test constants
SESSION_ID = "python_chrome_to_safari_test"
USER_ID = "browser_interop_user"

# Debug mode - set to True to see browser UI
DEBUG = False

# Redis configuration
REDIS_CONFIG = {
    "redis_url": "redis://localhost:6379/0",
    "key_prefix": "browserstate"
}

# Path to the test HTML file
TEST_HTML_PATH = Path(__file__).parent.parent.parent.parent / "typescript" / "examples" / "shared" / "test.html"
TEST_URL = f"file://{TEST_HTML_PATH.absolute()}"

def fail_test(message):
    """Fail the test with a clear error message."""
    print(f"\n❌ TEST FAILED: {message}")
    sys.exit(1)

async def create_browser_state():
    """Create browser state in Chrome and store it in Redis."""
    print("🚀 Starting Chrome State Creation using Python\n")

    try:
        # Initialize BrowserState with Redis storage
        print("🔧 Creating BrowserState with Redis storage...")
        options = BrowserStateOptions(
            user_id=USER_ID,
            redis_options=REDIS_CONFIG
        )
        browser_state = BrowserState(options)

        # List existing sessions
        print("\n📋 Listing existing sessions...")
        sessions = browser_state.list_sessions()
        print(f"Found {len(sessions)} session(s): {', '.join(sessions) if sessions else 'None'}")

        # Mount the session
        print(f"\n📥 Mounting session: {SESSION_ID}")
        mount_result = browser_state.mount_session(SESSION_ID)
        user_data_dir = mount_result["path"]
        print(f"📂 Mounted at: {user_data_dir}")

        # Launch Chrome browser with the mounted state
        print("\n🌐 Launching Chrome browser with Playwright...")
        async with async_playwright() as p:
            browser = await p.chromium.launch_persistent_context(
                user_data_dir=user_data_dir,
                headless=not DEBUG
            )

            try:
                # Create a new page
                page = await browser.new_page()

                # Navigate to the test HTML page
                print(f"\n📄 Loading test page: {TEST_URL}")
                await page.goto(TEST_URL)
                print("✅ Test page loaded")
                
                # Clear existing localStorage to start fresh
                await page.evaluate("() => { localStorage.clear(); }")
                
                # Wait for the page to load completely
                await page.wait_for_timeout(1000)

                # Add Chrome-specific test data to localStorage
                print("\n📝 Adding Chrome-specific test data to localStorage...")
                
                # Add notes to localStorage
                test_notes = [
                    "Python-Chrome created note 1",
                    "Python-Chrome created note 2",
                    "Python-Chrome created note 3"
                ]

                for note in test_notes:
                    await page.fill("#noteInput", note)
                    await page.click("button:text('Add Note')")
                    await page.wait_for_timeout(500)  # Wait for animation

                # Add browser metadata to localStorage
                await page.evaluate("""() => {
                    localStorage.setItem('browserMetadata', JSON.stringify({
                        browser: 'Chrome',
                        createdBy: 'Python',
                        timestamp: new Date().toISOString(),
                        userAgent: navigator.userAgent
                    }));
                }""")

                # Verify notes were added
                notes_count = await page.evaluate("""() => {
                    return JSON.parse(localStorage.getItem('notes') || '[]').length;
                }""")
                
                # Verify notes were added successfully
                if notes_count != len(test_notes):
                    fail_test(f"Expected {len(test_notes)} notes, but found {notes_count}")
                
                print(f"✅ Added {notes_count} notes")

                # Get the notes data for verification
                notes_data = await page.evaluate("""() => {
                    return localStorage.getItem('notes');
                }""")
                
                # Get browser metadata
                metadata_json = await page.evaluate("""() => {
                    return localStorage.getItem('browserMetadata');
                }""")
                
                # Verify data is not empty
                if not notes_data or not metadata_json:
                    fail_test("Test data is empty")
                
                print(f"📝 Notes data: {notes_data}")
                print(f"📝 Browser metadata: {metadata_json}")
                
                # Wait to ensure data is properly saved
                await page.wait_for_timeout(1000)

            finally:
                await browser.close()

        # Unmount the session
        print("\n🔒 Unmounting session...")
        browser_state.unmount_session()
        print("✅ Session unmounted")

        print("\n✨ Chrome state creation complete!")
        print("Now run verify_state.py to verify the state in Safari")

    except Exception as e:
        fail_test(f"Error during Chrome state creation: {str(e)}")

if __name__ == "__main__":
    asyncio.run(create_browser_state()) 